pkg/kubevirt/utils.go

```diff
@@ -138,36 +153,86 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
 	template.Spec = *ctx.KubevirtMachine.Spec.VirtualMachineTemplate.Spec.Template.Spec.DeepCopy()
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template)
+	switch cloudInitType {
+	case cloudInitNoCloud:
+		err := mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
+		if err != nil {
+			return nil, err
+		}
+		return template, nil
```

Comment:
I don't think there is any value in the mergo logic here. There's only one possible solution for how the CloudIniitNoCloud volume source can be configured. All we need to do is preserve the volume name, and override whatever is in the CloudInitNoCloud volume source. Anything else can result in an unusable configuration.

This also removes the need to perform any error checking in this function.

Reply:
@davidvossel, this is the simpler method than doing a manual three-way merge.
The logic right now is, if there is a volume called "cloudinitvolume" with a non-nil cloudInitNoCloud volume source it will merge the existing one, which has whatever properties configured, with the one setting the same for the bootstrap secret name.

I'm happy to update if you are really sure. The team is requiring this feature as soon as it's possible.

Reply:
This should be all we need.

```
if index >= 0 {
    template.Spec.Volumes[index] = cloudInitVolume
} else {
    template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
}
```

Reply:
I'm unsure whether that would work, the logic is stating to have either a volume added or a disk.

The current way things work means that
- you can bring your custom cloudInitNoCloud and we'll provide the secret, so that current behaviour is preserved but you can optionally use cloudInitNoCloud
  - e.g: providing custom noCloud networkData
- a cloudInit disk will be only added if one doesn't already exist, so that you can specify the type of disk that it is (if I wanted CDROM instead)

and i think that is useful

Reply:
i'm not following how the `mergo` usage is necessary to achieve what you previously replied with. I just recommending that the volume be overwritten if the index is known, or appended otherwise.